src/k2/persistence/plog_service/PlogServer.cpp

```diff
@@ -0,0 +1,161 @@
+    }
+    _plogMap.insert(std::pair<String,PlogPage >(std::move(request.plogId), PlogPage()));
+    return RPCResponse(Statuses::S201_Created("plog created"), dto::PlogCreateResponse());
+};
+
+seastar::future<std::tuple<Status, dto::PlogAppendResponse>>
+PlogServer::_handleAppend(dto::PlogAppendRequest&& request){
+    K2DEBUG("Received append request for " << request.plogId << " with size" << request.payload.getSize() << " and offset " << request.offset);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), dto::PlogAppendResponse());
+    }
+    if (iter->second.sealed){
+         return RPCResponse(Statuses::S409_Conflict("plog is sealed"), dto::PlogAppendResponse());
+    }
+    if (iter->second.offset != request.offset){
+        return RPCResponse(Statuses::S400_Bad_Request("offset inconsistent"), dto::PlogAppendResponse());
+    }
+    if (iter->second.offset + request.payload.getSize() > PLOG_MAX_SIZE){
+         return RPCResponse(Statuses::S400_Bad_Request("exceeds pLog limit"), dto::PlogAppendResponse());
+    }
+
+    dto::PlogAppendResponse response;
+    response.offset = iter->second.offset + request.payload.getSize();
+    response.bytes_appended = request.payload.getSize();
+
+    iter->second.offset += request.payload.getSize();
+    iter->second.payload.copyFromPayload(request.payload, request.payload.getSize());
+    
+    return RPCResponse(Statuses::S200_OK("append scuccess"), std::move(response));
+};
+
+
+seastar::future<std::tuple<Status, dto::PlogReadResponse>>
+PlogServer::_handleRead(dto::PlogReadRequest&& request){
+    K2DEBUG("Received read request for " << request.plogId << " with offset " << request.offset);
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), dto::PlogReadResponse());
+    }
+    if (iter->second.offset < request.offset + request.size){
+         return RPCResponse(Statuses::S400_Bad_Request("exceed the maximun length"), dto::PlogReadResponse());
+    }
+    
+    iter->second.payload.seek(request.offset);
+    dto::PlogReadResponse response{.payload=iter->second.payload.share(request.size)};
+    return RPCResponse(Statuses::S200_OK("read success"), std::move(response));
+};
+
+
+seastar::future<std::tuple<Status, dto::PlogSealResponse>>
+PlogServer::_handleSeal(dto::PlogSealRequest&& request){
+    K2DEBUG("Received seal request for " << request.plogId);
+    dto::PlogSealResponse response;
+    auto iter = _plogMap.find(request.plogId);
+    if (iter == _plogMap.end()) {
+        return RPCResponse(Statuses::S404_Not_Found("plog does not exist"), std::move(response));
+    }
+    if (iter->second.sealed){
+        response.offset = iter->second.offset;
+        return RPCResponse(Statuses::S409_Conflict("plog already sealed"), std::move(response));

```

Comment:
I think sealing should be an immutable operation - if a plog has been sealed, it should be ok to seal again

Reply:
changed the seal to be an immutable operation